interpreter/interpreter.go

```diff
@@ -38,22 +38,96 @@ type Interpreter interface {
 		decorators ...InterpretableDecorator) (Interpretable, error)
 }
 
+// EvalObserver is a functional interface that accepts an expression id and an observed value.
+type EvalObserver func(id int64, programStep interface{}, value ref.Val)
+
+func Observe(observers ...EvalObserver) InterpretableDecorator {
+	if len(observers) == 1 {
+		return decObserveEval(observers[0])
+	}
+	observeFn := func(id int64, programStep interface{}, val ref.Val) {
+		for _, observer := range observers {
+			observer(id, programStep, val)
+		}
+	}
+	return decObserveEval(observeFn)
+}
+
+// EvalCanceledError represents a cancelled program evaluation operation.
+type EvalCanceledError struct {
+	// Type identifies the cause of the cancellation.
+	Cause CancelationCause
+}
+
+func (e EvalCanceledError) Error() string {
+	switch e.Cause {
+	case ContextCancelled:
+		return "operation cancelled: context was cancelled"
```

Comment:
Consider adding a string message to the `EvalCancelledError` rather than using a switch here, and then use a statically declared variable for the predefined error with cause where you need to.